File: container_imgs/data_uploader/f_data_uploader/sql/predictions.py
from psycopg2.extras import execute_values, execute_batch

from f_data_uploader.database import db
from f_data_uploader.logger import logger


def get_matchday_points(matchday: dict) -> list[str]:
    with db.get_cursor() as cur:
        cur.execute(
            """
            WITH col_points as (
                select user_id, season, matchday, col_num, count(*) as points
                from bavariada.predictions
                where season = %s and matchday = %s and is_correct = true
                group by user_id, season, matchday, col_num
            )

            SELECT user_id, season, matchday, max(points) as points
            FROM col_points
            GROUP BY user_id, season, matchday;
            """,
            (
                matchday["season"],
                matchday["matchday"],
            ),
        )

        columns = [desc[0] for desc in cur.description]
        rows = cur.fetchall()

        matchday_points = [
            {columns[i]: value for i, value in enumerate(row)} for row in rows
        ]

    return matchday_points


def insert_predictions(users_predictions: list[dict]):
    with db.get_cursor() as cur:
        query = """
            INSERT INTO bavariada.predictions (user_id, season, matchday, col_num, match_num, prediction, is_elige8)
            VALUES %s
            ON CONFLICT (user_id, season, matchday, col_num, match_num)
            DO UPDATE SET
                prediction = EXCLUDED.prediction,
                is_elige8 = EXCLUDED.is_elige8
        """
        values = [
            (
                user_predictions["user_id"],
                user_predictions["season"],
                user_predictions["matchday"],
                user_predictions["col_num"],
                user_predictions["match_num"],
                user_predictions["prediction"],
                user_predictions["is_elige8"],
            )
            for user_predictions in users_predictions
        ]

        execute_values(cur, query, values)

        uploaded_infos = {
            (
                user_predictions["user_id"],
                user_predictions["season"],
                user_predictions["matchday"],
            )
            for user_predictions in users_predictions
        }
        for uploaded_info in uploaded_infos:
            logger.info(
                f"Successfully upserted predictions for user {uploaded_info[0]}"
                f" in season {uploaded_info[1]} matchday {uploaded_info[2]}"
            )


def update_predictions(matches: list[dict]):
    with db.get_cursor() as cur:
        query = """
            UPDATE bavariada.predictions
            SET is_correct = (prediction = %s)
            WHERE season = %s AND matchday = %s AND match_num = %s
        """
        values = [
            (
                match["result"],
                match["season"],
                match["matchday"],
                match["match_num"],
            )
            for match in matches
        ]

        execute_batch(cur, query, values)

File: container_imgs/app/internal/db/predictions.go
package db

import (
	"github.com/jackc/pgx/v5/pgtype"
)

type UserPredictions struct {
	UserName    string          `json:"user_name"`
	Predictions [][]string      `json:"predictions"`
	IsCorrect   [][]pgtype.Bool `json:"is_correct"`
	IsElige8    [][]pgtype.Bool `json:"is_elige8"`
}

func GetUserPredictions(matchday int32) (map[string]UserPredictions, error) {
	db := New()

	rows, err := db.Query(
		`WITH predictions_per_user_match AS (
			SELECT
				user_id,
				match_num,
				array_agg(prediction ORDER BY col_num) as predictions,
				array_agg(is_correct ORDER BY col_num) as is_correct,
				array_agg(is_elige8 ORDER BY col_num) as is_elige8
			FROM bavariada.predictions
			WHERE season = '2024-2025' AND matchday = $1
			GROUP BY user_id, match_num
		)
		SELECT
			users.id as user_id,
			COALESCE(COALESCE(users.raw_user_meta_data->>'display_name', users.email), users.email) AS display_name,
			array_agg(predictions_per_user_match.predictions ORDER BY predictions_per_user_match.match_num) as predictions,
			array_agg(predictions_per_user_match.is_correct ORDER BY predictions_per_user_match.match_num) as is_correct,
			array_agg(predictions_per_user_match.is_elige8 ORDER BY predictions_per_user_match.match_num) as is_elige8
		FROM predictions_per_user_match
		LEFT JOIN auth.users users ON predictions_per_user_match.user_id = users.id
		GROUP BY users.id, COALESCE(COALESCE(users.raw_user_meta_data->>'display_name', users.email), users.email);`,
		matchday,
	)
	if err != nil {
		return map[string]UserPredictions{}, err
	}
	defer rows.Close()

	usersPredictions := map[string]UserPredictions{}
	for rows.Next() {
		userPredictions := UserPredictions{}
		var userId string
		err = rows.Scan(
			&userId,
			&userPredictions.UserName,
			&userPredictions.Predictions,
			&userPredictions.IsCorrect,
			&userPredictions.IsElige8,
		)
		if err != nil {
			return map[string]UserPredictions{}, err
		}
		usersPredictions[userId] = userPredictions
	}

	return usersPredictions, nil
}
